pyomo/core/expr/numvalue.py

```diff
@@ -180,18 +180,12 @@ def value(obj, exception=True):
     """
     if obj.__class__ in native_types:
         return obj
+    if obj.__class__ is NumericConstant:
+        if obj.value is None:
+            raise ValueError("No value for uninitialized NumericConstant object %s" % (obj.name,))
+        return obj.value
     try:
-        numeric = obj.as_numeric()
-    except AttributeError:
-        try:
-            numeric = as_numeric(obj)
-        except ValueError:
-            if isinstance( obj, string_types + (text_type, binary_type) ):
-                native_types.add(type(obj))
-                return obj
-            raise
-    try:
-        tmp = numeric(exception=exception)
+        tmp = obj(exception=exception)
```

Comment:
Can you give an example?

Reply:
```
m.p = my_numeric_like_class_that_pyomo_has_never_seen_before()
#...
def some_rule(m, i):
    if i == value(m.p):  # because we train users to wrap things in value() calls...
        return Constraint.Skip
    #...
```
In this case, the first time Pyomo sees this class is in the `value()` call -- not in as_numeric()

Reply:
In addition, as with `value()` below, the reason to use the `.as_numeric()` attribute is so that the nominal case works quickly (Pyomo numeric types all have `as_numeric` attributes)  The way this PR is written, you will get a generic exception for anything that is not callable with an `exception` argument - either because it was a Pyomo numeric type that failed to evaluate, or if it was a non-numeric type.